### DISCUSSION TITLE
is it possible to convert a texture with a "rgba8unorm" format to "bgra8unorm" ?

### DISCUSSION THREAD

**Discussion by tlecoz** [gpuweb/gpuweb#4082] [Q&A] [answered]
I want to use a texture generated by a renderPipeline as input of a computeShader. 

It works but I need to use a texture with a GPUTextureUsage.STORAGE_BINDING and this usage doesn't work with "bgra8unorm" which is the value of navigator.gpu.getPreferredCanvasFormat()

so it imply that my renderPipeline also use a colorFormat defined to "rgba8unorm" and other textures too.... 

I would like to remove this particular case and convert my texture in bgra8unorm between the computePipeline pass and the renderPipeline pass but I don't even know if it's possible.

Is it possible ?

**Comment by kainino0x** [comment, marked as answer]
- If your system supports bgra8unorm-storage you can enable that and then you **can** use it as storage (I think chrome on mac/windows always supports this but I'm not positive)
- The preferred format is only for when you're using a canvas (the format in ctx.configure()), in particular using it for display. You can use rgba8unorm for rendering without penalty in the middle of your pipeline; the penalty comes when it's displayed.
    - EDIT: additional clarification, it's also (at least in principle) fine and good to use rgba8unorm for a canvas when you're using it only to interop with other APIs like 2d canvas / webgl / video encoding / etc. In this case you would typically want to use `OffscreenCanvas`.
- It's better to make the browser deal with a non-preferred format than it is to add a copy in your application. If you do a copy in your code there will **always** be a copy, whereas the browser **may or may not** need a copy to implement non-preferred formats, which is more optimal.
- If you *still* need to do a conversion between two different formats, the way to do it is with a render pass: draw a full-screen quad that reads from the source texture, and render it into the destination texture.

**Reply by tlecoz** [reply]
Thank you for your message and your reactivity !
